File: sdks/tokenizer-sdk/src/accounts/emergencyRecord.ts
/**
 * EmergencyRecord account deserializer.
 *
 * #[repr(C)] layout (160 bytes):
 *   0: account_key (u8)
 *   1: version (u8)
 *   2: asset ([u8;32])
 *  34: old_asset_token ([u8;32])
 *  66: old_owner ([u8;32])
 *  98: recovery_type (u8)
 * 104: created_at (i64)        — padded from 99
 * 112: bump (u8)
 * 113: reason (u8)
 * 120: shares_transferred (u64) — padded from 114
 * 128: remainder_token ([u8;32])
 */
import { type Address, getStructDecoder, transformDecoder } from "gill";
import { AccountKey } from "../constants.js";
import { addr, i64d, pad, u8d, u64d } from "./decode.js";

export const EMERGENCY_RECORD_SIZE = 160;
export const EMERGENCY_RECORD_OFFSET_ASSET = 2;

export interface EmergencyRecord {
  accountKey: number;
  version: number;
  asset: Address;
  oldAssetToken: Address;
  oldOwner: Address;
  recoveryType: number;
  createdAt: bigint;
  bump: number;
  reason: number;
  sharesTransferred: bigint;
  remainderToken: Address;
}

const rawDecoder = getStructDecoder([
  ["accountKey", u8d],
  ["version", u8d],
  ["asset", addr],
  ["oldAssetToken", addr],
  ["oldOwner", addr],
  ["recoveryType", u8d],
  ["_p0", pad(5)],
  ["createdAt", i64d],
  ["bump", u8d],
  ["reason", u8d],
  ["_p1", pad(6)],
  ["sharesTransferred", u64d],
  ["remainderToken", addr],
]);

export const emergencyRecordDecoder = transformDecoder(rawDecoder, ({ _p0, _p1, ...rest }) => rest);

export function decodeEmergencyRecord(data: Uint8Array): EmergencyRecord {
  if (data.length < EMERGENCY_RECORD_SIZE) {
    throw new Error(`EmergencyRecord: expected ${EMERGENCY_RECORD_SIZE} bytes, got ${data.length}`);
  }
  const accountKey = data[0];
  if (accountKey !== AccountKey.EmergencyRecord) {
    throw new Error(`EmergencyRecord: invalid account key ${accountKey}`);
  }
  return emergencyRecordDecoder.decode(data);
}

File: programs/tokenizer/src/state/emergency_record.rs
use pinocchio::Address;

use super::EMERGENCY_RECORD_SEED;

/// Audit trail for emergency token recovery operations.
/// Prevents double recovery and provides on-chain proof.
/// PDA: ["emergency_record", old_asset_token.key()]
#[repr(C)]
pub struct EmergencyRecord {
    pub account_key: u8,           // AccountKey::EmergencyRecord (10)
    pub version: u8,
    pub asset: [u8; 32],           // Parent asset
    pub old_asset_token: [u8; 32], // The recovered token
    pub old_owner: [u8; 32],       // Previous owner (lost wallet)
    pub recovery_type: u8,         // 0 = burn_and_remint, 1 = split_and_remint
    pub created_at: i64,
    pub bump: u8,

    // Terms & Conditions fields

    /// Recovery reason (RecoveryReason enum). 0 = LostKeys (default).
    pub reason: u8,

    /// Number of shares transferred (for partial burn_and_remint). 0 = full transfer.
    pub shares_transferred: u64,

    /// Remainder token PDA address (for partial burn_and_remint). Zero = no remainder.
    pub remainder_token: [u8; 32],
}

impl EmergencyRecord {
    pub const LEN: usize = core::mem::size_of::<Self>();

    #[inline(always)]
    pub unsafe fn load(data: &[u8]) -> &Self {
        &*(data.as_ptr() as *const Self)
    }

    #[inline(always)]
    pub unsafe fn load_mut(data: &mut [u8]) -> &mut Self {
        &mut *(data.as_mut_ptr() as *mut Self)
    }

    pub fn derive_pda(
        old_asset_token: &[u8; 32],
        program_id: &Address,
    ) -> (Address, u8) {
        Address::find_program_address(
            &[EMERGENCY_RECORD_SEED, old_asset_token],
            program_id,
        )
    }
}
